x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalMathProcessor.java

```diff
@@ -24,13 +26,78 @@
     public enum BinaryOptionalMathOperation implements BiFunction<Number, Number, Number> {
 
         ROUND((l, r) -> {
-            double tenAtScale = Math.pow(10., r.longValue());
+            long rLong = r.longValue();
+            if (l instanceof Long || l instanceof Integer) {
+                long lLong = l.longValue();
+                if (lLong == 0L || rLong >= 0) {
+                    return lLong;
+                }
+
+                long digitsToRound = -rLong;
+                int digits = (int) (Math.log10(Math.abs(l.doubleValue())) + 1);
+                if (digits <= digitsToRound) {
+                    return 0L;
+                }
+
+                long divider = (long) Math.pow(10, digitsToRound);
+                long middleResult = lLong / divider;
+                long remainder = lLong % divider;
+                if (remainder >= 5 * (long) Math.pow(10, digitsToRound - 1)) {
+                    middleResult++;
+                } else if (remainder <= -5 * (long) Math.pow(10, digitsToRound - 1)) {
+                    middleResult--;
+                }
+
+                long result = middleResult * divider;
+                if (Long.signum(result) == Long.signum(lLong)) {
+                    return result;
+                }// otherwise there was an overflow on long values, fall back to floating point implementation.
```

Comment:
the `ROUND` docs state that "The returned numeric data type is the same as the data type of numeric_exp." This property does no longer hold with the fallback to double arithmetic.

Reply:
Yeah, it didn't hold even before this change (it always returned double), but it's a good moment to fix it 👍